### PR TITLE
Custom labels

### PR DESCRIPTION
**Pull request by Alkarex**
https://github.com/FreshRSS/FreshRSS/issues/928
https://github.com/FreshRSS/FreshRSS/issues/1367

### CODE REVIEW

**Comment by Alkarex** [comment]
First draft

**Comment by Alkarex** [comment]
image

**Comment by Alkarex** [comment]
image

File: app/Models/EntryDAO.php [hunk 1]
```diff
@@ -129,13 +129,18 @@ protected function autoUpdateDb($errorInfo) {
 						return $this->addColumn($column);
 					}
 				}
-			} elseif ($errorInfo[0] === '42S02' && stripos($errorInfo[2], 'entrytmp') !== false) {	//ER_BAD_TABLE_ERROR
-				return $this->createEntryTempTable();	//v1.7
+			} elseif ($errorInfo[0] === '42S02') {	//ER_BAD_TABLE_ERROR
```
Comment (aledeg) [inline app/Models/EntryDAO.php]: I think that the magic value 42S02 should be in class constant.

File: app/Models/EntryDAOPGSQL.php [hunk 2]
```diff
@@ -12,8 +12,13 @@ public function sqlHexEncode($x) {
 
 	protected function autoUpdateDb($errorInfo) {
 		if (isset($errorInfo[0])) {
-			if ($errorInfo[0] === '42P01' && stripos($errorInfo[2], 'entrytmp') !== false) {	//undefined_table
-				return $this->createEntryTempTable();
+			if ($errorInfo[0] === '42P01') {	//undefined_table
```
Comment (aledeg) [inline app/Models/EntryDAOPGSQL.php]: Same here

File: app/Models/EntryDAO.php (hunk 1 above)
Comment (Alkarex) [reply app/Models/EntryDAO.php]: Done in https://github.com/FreshRSS/FreshRSS/pull/2027/commits/036be54ac70b707c1113788a6f9e4146f9981592

File: app/Models/EntryDAOPGSQL.php (hunk 2 above)
Comment (Alkarex) [reply app/Models/EntryDAOPGSQL.php]: Done in https://github.com/FreshRSS/FreshRSS/pull/2027/commits/036be54ac70b707c1113788a6f9e4146f9981592

**Comment by Alkarex** [comment]
Now with the possibility to add new tags:

image

**Comment by Alkarex** [comment]
If someone tried this branch, the last commit introduces a new field. Either drop the tag tables (which will be re-created automatically), or add the field with:

```sql
ALTER TABLE freshrss_alice_tag ADD COLUMN attributes TEXT;
```

File: app/Models/CategoryDAO.php [hunk 3]
```diff
@@ -9,7 +9,7 @@ public function addCategory($valuesTmp) {
 		$stm = $this->bd->prepare($sql);
 
 		$values = array(
-			mb_strcut($valuesTmp['name'], 0, 255, 'UTF-8'),
+			mb_strcut($valuesTmp['name'], 0, 191, 'UTF-8'),
```
Comment (aledeg) [inline app/Models/CategoryDAO.php]: Not a big fan of magic values. It would be better to have constants. Same comment goes for all magic values.

**Review by aledeg — CHANGES_REQUESTED** [review]

File: app/Models/CategoryDAO.php (hunk 3 above)
Comment (Alkarex) [reply app/Models/CategoryDAO.php]: Done in https://github.com/FreshRSS/FreshRSS/pull/2027/commits/d8a3c8cfb90c97b72f549c7ba13870ebf03249b5 (but not a big fan of the additional complexity of a constant in this case :-) )

**Comment by Alkarex** [comment]
TODO:
* [x] Drop-down menu to e.g. remove tags
* [x] Mark as read
* [x] Read/unread views
* [x] Live-change unread tags when marking as read/unread
* [x] Keep tagged articles during purge?
* [x] Investigate tag options in API / mobile clients

**Review by aledeg — APPROVED** [review]
:+1: :tada:

**Comment by aledeg** [comment]
Nice work @Alkarex

**Comment by Alkarex** [comment]
Thanks @aledeg , I hope you will like it :-)

**Comment by Alkarex** [comment]
Google Reader API implemented. Fully functional with News+:

screenshot_20180924-232132_news

screenshot_20180924-232212_news

**Pull request merged**